include/manif/impl/dhu/DHu_base.h

```diff
@@ -0,0 +1,411 @@
+    "Argument does not inherit from DHuBase !");
+  
+  const auto& m_DHu = static_cast<const DHuBase<_DerivedOther>&>(m);
+
+  if (J_mc_ma)
+  {
+    (*J_mc_ma) = m_DHu.inverse().adj();
+  }
+
+  if (J_mc_mb)
+  {
+    J_mc_mb->setIdentity();
+  }
+
+  return LieGroup(real()*m_DHu.real(),
+                  Quaternion((real()*m_DHu.dual()).coeffs()+(dual()*m_DHu.real()).coeffs()));
+}
+
+template <typename _Derived>
+template <typename _EigenDerived>
+Eigen::Matrix<typename DHuBase<_Derived>::Scalar, 3, 1>
+DHuBase<_Derived>::act(const Eigen::MatrixBase<_EigenDerived> &v,
+                       tl::optional<Eigen::Ref<Eigen::Matrix<Scalar, 3, 6>>> J_vout_m,
+                       tl::optional<Eigen::Ref<Eigen::Matrix<Scalar, 3, 3>>> J_vout_v) const
+{
+  assert_vector_dim(v, 3);
+  const Rotation R(rotation());
+
+  if (J_vout_m)
+  {
+    // @todo
+  }
+
+  if (J_vout_v)
+  {
+    // @todo
+  }
+
+  // @todo do it in a dual quat fashion
+  return translation() + R * v;
+}
+
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Jacobian
+DHuBase<_Derived>::adj() const
+{
+  // @todo
+
+  return Jacobian();
+}
+
+// DHu specific function
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Transformation
+DHuBase<_Derived>::transform() const
+{
+  Transformation T = Transformation::Identity();
+  T.template topLeftCorner<3,3>()  = rotation();
+  T.template topRightCorner<3,1>() = translation();
+  return T;
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Isometry
+DHuBase<_Derived>::isometry() const
+{
+  return Isometry(transform());
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Rotation
+DHuBase<_Derived>::rotation() const
+{
+  return real().matrix();
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Real
+DHuBase<_Derived>::real() const
+{
+  return Real(coeffs().template head<4>());
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Dual
+DHuBase<_Derived>::dual() const
+{
+  return Real(coeffs().template tail<4>());
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::Translation
+DHuBase<_Derived>::translation() const
+{
+  return dual()*real().conjugate()*Scalar(2);
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::LieGroup
+DHuBase<_Derived>::conjugate() const
+{
+  return LieGroup(real().conjugate(), dual().conjugate());
+}
+
+template <typename _Derived>
+typename DHuBase<_Derived>::LieGroup
+DHuBase<_Derived>::conjugatedual() const
+{
+  return LieGroup((DataType() << coeffs().template head<4>(), 
+                                -coeffs().template tail<4>()).finished());
+}
+
+template <typename _Derived>
+void DHuBase<_Derived>::normalize()
+{
+  const Scalar n = coeffs().template head<4>().norm();
+  coeffs() /= n;
+  // real and dual parts are orthogonal
+  coeffs().template tail<4>() -= (coeffs().template head<4>().dot(coeffs().template tail<4>())) * coeffs().template head<4>();
+}
```

Comment:
I did not understand this part in the paper. What are we to normalize? The rotational quaternion (4 components) and then the translational quaternion by the same scaling (4 more components)? Or all the real+dual quaterninon (8 components)? Or both the real and dual normalized separately?

That is, at the end of normalize(), what should we have:

1) | Real | = 1 and | Dual | = anything (just as in the code)
2) | (Real,Dual) | = 1
3) | Real | = 1 AND | Dual | = 1

??

Reply:
In [this source](http://web.cs.iastate.edu/~cs577/handouts/dual-quaternion.pdf) the author defines 3 conjugates (Sec.5) and more specifically the second conjugate that "follows from the classical quaternion conjugation": `σ∗ = p∗ + ϵq∗`. Sec.6 states that "a dual quaternion is unit if σ ⊗ σ∗ = 1" which implies that the real part `p` must be a unit quaternion and the dual part `p` must be orthogonal to the real part `p`. So at the end of `normalize()` we should have:

4.a |Real| = 1
4.b Real . Dual = 0

These are the conditions I'm checking for at construction (see [here](https://github.com/artivis/manif/pull/165/commits/b9f046007476b3ca72679445c942d2aca99f9e9c#diff-f72abf273a625bdfe02f36e08d37c8daR396)). And I'm testing `normalize()` in [this test](https://github.com/artivis/manif/pull/165/commits/b9f046007476b3ca72679445c942d2aca99f9e9c#diff-a2ad8132fa38f63df0d82a193d4e92b7R132) which seems to pass.

Reply:
How is the product in 4b defined? Is it regular quaternion product?

Reply:
Sec.6 Eq.21 of the aforementioned reference: `p0q0 + p1q1 + p2q2 + p3q3 = 0`

Reply:
OK I see, so they are orthogonal as 4-vectors: p.tr * q = 0